hw1/matrices.py

```diff
@@ -0,0 +1,66 @@
+from math import sqrt, acos, isclose, pi
+from typing import List
+
+
+def dot_product(vector1, vector2):
+    if len(vector1) != len(vector2):
+        raise IndexError(
+            f"Cannot perform dot product of vectors of different dimensions: {len(vector1)} and {len(vector2)} "
+        )
+    return sum((x * y for x, y in zip(vector1, vector2)))
+
+
+def magnitude(vector):
+    return sqrt(dot_product(vector, vector))
+
+
+def angle_between_vectors(vector1, vector2):
+    return acos(dot_product(vector1, vector2) / (magnitude(vector1) * magnitude(vector2)))
+
+
+def check_and_matrix_shape(matrix):
+    if len(matrix) == 0 or len(matrix[0]) == 0 or any((len(row) != len(matrix[0]) for row in matrix)):
+        raise IndexError("Matrix cannot be empty or contain empty rows")
+    if any((len(row) != len(matrix[0]) for row in matrix)):
```

Comment:
In these two checks you are repeating the same assertion. Try to write the first one a bit differently, so that you don't check for the same thing twice.